### PR TITLE
`Development`: Add API versioning

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I documented the Java code using JavaDoc style.

### Motivation and Context

Especially with the usage of Apps in mind, changing the API of the application becomes a complex endeavour as you have to synchronise the changes in app clients and the application itself. Users with an old version of the app will suddenly face errors.

We want to be able to add versions to the API to avoid such issues.

### Description

I implemented the versioning framework in the `versioning` with additional tests. Check the documentation and the JavaDocs for detailed explanations. Because we add the `/api` prefix through the new mapper, I had to remove the corresponding RestMapping annotations throughout the system.

Every versioned endpoint, however, has the empty version mapped to the latest version too (/api/v1/... == /api/... for v1 latest).

To ensure that all environments get tested, I added an abstract method to the abstract integration test, which should be implemented in all direct subclasses (i.e. test environments). If it isn't, there will be build errors. I have not found a better solution. If you have one, please feel free to let me know.

Finally, I switched to another opendocs implementation that is better supported. However, some issues remain, so it's deactivated by default. We don't use it right now anyway. In a follow-up, I aim to get this running in some form.

**Addendum**:  Because we only want to activate versioning for verified REST conform endpoints, I added an annotation to whitelist endpoints for versioning. Once all endpoints were checked, which can take some time, we can remove this annotation and checks for it.

I bumped the coverage checks because due to the whitelisting (and since we don't have any whitelisted endpoints yet), a section doesn't get covered which will get covered once we whitelist our first endpoint.

### Steps for Testing

Log in and go through some pages. Make sure everything is working as expected. The major load will be code review here. This is possible because the generation of the paths is automated (if it works for one, it works for all).

### Steps for Code Review
1. Review the versioning framework
    - main: config/VersioningConfiguration.java
    - main: config/RequestMappingConfiguration.java
    - main: exception/...
    - main: versioning/...
    - test: versioning/...
2. Review the rest
    - Make sure that no things got removed that should not be removed
    - Make sure that all endpoints that were previously available are available now too (according to the annotations)
    - Make sure that all annotations declaring the `/api` prefix got removed, otherwise this would turn into a 404.

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| RequestMappingConfiguration.java | 100% |                            |
| VersioningConfiguration.java | 100% |               |
| VersionRangeComparator.java | 94.3% | ✅                           |
| VersionRangeFactory.java | 91.6% | ✅                           |
| VersionRangesRequestCondition.java | 86.2% | ✅                           |
| VersionRequestMappingHandlerMapping.java | 90.0% | ✅              |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:186]: Should we expect more of these duplicates with and without `version`? We could also use an option in the regex like this: `/api/({version:v\\d+}/)?public/**`, can't we?

File: docs/dev/api-versioning.rst
Comment (JohannesStoehr) [inline docs/dev/api-versioning.rst:11]: What is the use case for this? We support versions 1-3 but not version 4 but from version 5 onwards we again support this?

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java:21]: Why did you use an array instead of two separate values?

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:57]: As far as I can tell, it is not specified what `result < 1` is supposed to mean for Integer.compare, see: https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Integer.html#compare(int,int)

File: src/main/java/de/tum/in/www1/artemis/web/rest/AeolusTemplateResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AeolusTemplateResource.java:28]: Is it possible to add another/a newly created annotation to keep the "aeolus" here?

File: src/main/java/de/tum/in/www1/artemis/web/rest/PlantUmlResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/PlantUmlResource.java:43]: Was that removal intentional?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: docs/dev/api-versioning.rst
Comment (julian-christl) [reply docs/dev/api-versioning.rst:11]: The change could be reverted because we see that it will conflict with another design decision in the future. However, the app already uses the new endpoint.

Probably unlikely, but I saw no reason to limit the functionality here. We don't have to use the feature as you can directly annotate the endpoint with a @VersionRange annotation.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java:21]: To avoid always having the `abcd = ` prefix.

I could also set value as an alias for `start` and only if there is an end version, you'd have to specify it. Would you prefer this?

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:57]: Not sure what you mean by that.
`the value 0 if x == y; a value less than 0 if x < y; and a value greater than 0 if x > y`

=> result < 1 <=> x "<=" y

File: src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:186]: No, I don't expect more duplicates. The rest is covered by the last catch-all request matcher.

I don't think that would work. The curly brackets mark the regex. In your example, you use the regex outside of the brackets would become part of the path matching, not the regex, same goes for the brackets.

`/api/{version:(v\\d+/)?}public/**` might work, but honestly, it gets confusing. Having two of these makes it pretty clear that one is for the default option and one is for the versioned options.

File: src/main/java/de/tum/in/www1/artemis/web/rest/PlantUmlResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/PlantUmlResource.java:43]: No, good catch

File: docs/dev/api-versioning.rst
Comment (JohannesStoehr) [reply docs/dev/api-versioning.rst:11]: Would there be any downside to simply extend the range in my example to 1-5 or from 1 onwards to infinity?
The reason against `@VersionRanges` would be more and more complicated code that has to be maintained

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java:21]: With specific values it would be a bit clearer and easier to understand what values can be used for the annotation. But I don't have hard feelings against the array

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:57]: Shouldn't it be
```suggestion
        if (result < 0) {
```

File: docs/dev/api-versioning.rst
Comment (julian-christl) [reply docs/dev/api-versioning.rst:11]: Yes, if you would simply extend the range, you would have to remove the "new/old/reverted" endpoint and suddenly everything using that endpoint would not be working anymore, but suddenly be confronted with a different functionality behind the same endpoint. Exactly the kind of thing we want to avoid with this PR in the first place.

Point 2: That is true. However, I'd argue that the maintenance is pretty low, if any. If there are significant changes in how Spring allows us to handle this situation - we would have to update everything anyway.

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:57]: No, see the comment in the body: "Range starts with or after the limit starts". In other words, if the Limit, which is A, starts afterwards, it doesn't include the range. If it starts with the range or before it includes the range. Hence, < 0 and eq 0 are relevant here.

Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:57]: But you don't check for < 0

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRange.java]: I changed it

File: src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/versioning/VersionRangeComparator.java:49]: Got it

**Review by JohannesStoehr — DISMISSED** [review]
Code

File: src/main/java/de/tum/in/www1/artemis/config/RequestMappingConfiguration.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/config/RequestMappingConfiguration.java:20]: Does a build agent even offer any API? Why do we need this for build agents as well?

File: src/test/java/de/tum/in/www1/artemis/AbstractArtemisIntegrationTest.java
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/AbstractArtemisIntegrationTest.java:245]: Can you add an ArchitectureTest for this? I can see that be overlooked if IntelliJ just offers a default implementation here.

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
server style (documentation) is failing.

File: src/main/java/de/tum/in/www1/artemis/config/RequestMappingConfiguration.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/config/RequestMappingConfiguration.java:20]: I want to avoid any issues. Maybe they do, maybe they don't. But I want this to be activated for all endpoints. If you don't have any, fine.
But build agents are the only profiles that get activated separately from core profiles.

**Review by BBesrour — DISMISSED** [review]
Tested on TS3. No errors encountered. Steps:
- Created programming exercise
- Checked result logs
- submit as student
- Made new announcement in course (received email)
- Private messages with another user

**Review by az108 — DISMISSED** [review]
Tested on TS3 and no errors occured:

- Navigated on various pages
- Created All types of Exercises
- Tested all types of exercises with sumbission
- Tested Communication Feture 
- Tested Course Management Features

**Review by BBesrour — APPROVED** [review]
re-approve

**Review by az108 — APPROVED** [review]
reapprove

**Review by JohannesStoehr — APPROVED** [review]
Reapprove code

**Review by laadvo — APPROVED** [review]
Code LGTM :+1:

**Review by laadvo — DISMISSED** [review]
Reapprove Code

**Review by BBesrour — APPROVED** [review]
re-approve

**Review by az108 — APPROVED** [review]
reapprove

**Review by JohannesStoehr — DISMISSED** [review]

**Review by Hialus — APPROVED** [review]

**Review by sarpsahinalp — DISMISSED** [review]

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove after merge

**Review by JohannesStoehr — APPROVED** [review]

**Review by sarpsahinalp — DISMISSED** [review]
Approve after changes

**Review by az108 — DISMISSED** [review]
Code LGTM after changes

**Review by laadvo — APPROVED** [review]
Reapprove code after merge conflict

**Pull request closed without merging**
